### ISSUE TITLE
[LoRA Phase 10.3] Implement Kernel Fusion Optimization

### ISSUE THREAD

**Issue by makr-code** [makr-code/ThemisDB#563]
## 📋 Description

**Priority**: Phase 10.3 - Priority #3 (Performance Optimization)  
**Prerequisites**: Phases 1-9 complete (all GPU backends functional)  
**Estimated Effort**: 2 weeks  
**Status Document**: `LORA_GPU_PHASE10_PLAN.md`

Implement kernel fusion to reduce memory bandwidth and kernel launch overhead. Fusing multiple operations into single kernels can provide 1.5-2x additional speedup on top of existing 50x improvement by reducing global memory traffic by 66-75%.

## 🎯 Goals

- [ ] Fuse forward pass operations into single kernel
- [ ] Fuse backward pass operations into single kernel
- [ ] Reduce memory bandwidth by 66-75%
- [ ] Achieve 1.5-2x additional speedup (3.2ms → 1.5-2.0ms per training step)
- [ ] Maintain numerical accuracy (< 1e-5 error)

### 1. Fused Forward Pass Kernel (CUDA)
- [ ] Design fused kernel: `fused_lora_forward(input, B, A, scaling, output)`
- [ ] Single kernel computes: `output = (input @ B) @ A * scaling`
- [ ] Use shared memory for intermediate results (h = input @ B)
- [ ] Avoid writing h to global memory
- [ ] Optimize for memory coalescing and bank conflicts

**Files**:
- `include/llm/lora_framework/cuda_fused_kernels.h` (new)
- `src/llm/lora_framework/kernels/cuda_fused_kernels.cu` (new)

**Expected Performance**:
- Current: 3 kernel launches (input@B, h@A, scale)
- Fused: 1 kernel launch
- Memory traffic reduction: 66%
- Speedup: 1.5-1.8x

### 2. Fused Backward Pass Kernel (CUDA)
- [ ] Design fused kernel: `fused_lora_backward(grad_out, input, B, A, scaling, grads)`
- [ ] Single kernel computes all gradients:
  - `grad_A = h^T @ grad_out * scaling`
  - `grad_B = input^T @ (grad_out @ A^T * scaling)`
  - `grad_input = (grad_out @ A^T) @ B^T * scaling`
- [ ] Use shared memory for intermediate computations
- [ ] Avoid multiple global memory passes
- [ ] Optimize thread block configuration

**Files**:
- Add to `cuda_fused_kernels.cu`

**Expected Performance**:
- Current: 4 kernel launches
- Fused: 1 kernel launch
- Memory traffic reduction: 75%
- Speedup: 1.7-2.0x

### 3. Fused Optimizer Update Kernel (CUDA)
- [ ] Fuse gradient accumulation + weight decay + momentum + update
- [ ] Single kernel: `fused_sgd_step(params, grads, momentum, lr, weight_decay)`
- [ ] Compute all updates in single pass over data
- [ ] Support both with/without momentum cases

**Files**:
- Add to `cuda_fused_kernels.cu`

**Expected Performance**:
- Current: 3-4 kernel launches
- Fused: 1 kernel launch
- Speedup: 1.3-1.5x

### 4. HIP Fused Kernels
- [ ] Port CUDA fused kernels to HIP
- [ ] Optimize for AMD GPU architecture (RDNA2/RDNA3)
- [ ] Use wave64 intrinsics where beneficial
- [ ] Test on AMD hardware

**Files**:
- `include/llm/lora_framework/hip_fused_kernels.h` (new)
- `src/llm/lora_framework/kernels/hip_fused_kernels.cpp` (new)

### 5. Vulkan Fused Shaders (Optional)
- [ ] Design fused compute shaders in GLSL
- [ ] Implement fused forward pass shader
- [ ] Implement fused backward pass shader
- [ ] Compile to SPIR-V and integrate

**Files**:
- `src/acceleration/vulkan/shaders/lora/fused_forward.comp` (new)
- `src/acceleration/vulkan/shaders/lora/fused_backward.comp` (new)

### 6. DirectX Fused Shaders (Optional)
- [ ] Design fused compute shaders in HLSL
- [ ] Implement fused forward pass shader
- [ ] Implement fused backward pass shader
- [ ] Compile and integrate with DirectX backend

**Files**:
- `src/acceleration/directx/shaders/lora/fused_forward.hlsl` (new)
- `src/acceleration/directx/shaders/lora/fused_backward.hlsl` (new)

### 7. Integration with GPULoRALayer
- [ ] Add fused kernel path to `GPULoRALayer::forward()`
- [ ] Add fused kernel path to `GPULoRALayer::backward()`
- [ ] Add fused optimizer update to `GPUSGDOptimizer::step()`
- [ ] Runtime selection: fused vs unfused (for debugging)
- [ ] Support both paths with compile-time flag

**Files**:
- `src/llm/lora_framework/gpu_lora_layers.cpp` (update)

### 8. Performance Analysis
- [ ] Benchmark fused vs unfused kernels
- [ ] Measure memory bandwidth reduction
- [ ] Profile with nsight/rocprof
- [ ] Analyze occupancy and register usage
- [ ] Compare across different tensor sizes

**Tools**:
- NVIDIA Nsight Compute
- AMD ROCm Profiler
- Custom benchmarking scripts

### 9. Testing and Validation
- [ ] Unit tests for fused kernels
- [ ] Numerical accuracy tests (fused vs unfused < 1e-5)
- [ ] Performance regression tests
- [ ] Test on different GPU architectures
- [ ] Validate gradient correctness

**Files**:
- `tests/test_fused_kernels.cpp` (new)
- Update `tests/test_gpu_lora_layers.cpp`

### 10. Documentation
- [ ] Document kernel fusion design and implementation
- [ ] Performance tuning guide for fused kernels
- [ ] Trade-offs: fused vs unfused
- [ ] Update Phase 10 plan with results

## ✅ Acceptance Criteria

- [ ] Fused forward pass kernel functional on CUDA/HIP
- [ ] Fused backward pass kernel functional on CUDA/HIP
- [ ] **Performance**: Training step 1.5-2.0ms (1.5-2x improvement)
- [ ] **Accuracy**: Numerical error < 1e-5 vs unfused kernels
- [ ] **Memory**: Bandwidth reduced by 66-75%
- [ ] All tests pass with fused kernels
- [ ] Gradient correctness verified
- [ ] Backward compatibility maintained (unfused path still works)

## 🔗 Dependencies

- CUDA 11.8+ (for CUDA kernels)
- ROCm 5.0+ (for HIP kernels)
- Existing GPU backends (CUDA, HIP functional)
- GPULoRALayer and GPUTensor (complete)
- Profiling tools (Nsight Compute, rocprof)

### Current Performance (Unfused)
| Operation | Time | Kernels |
|-----------|------|---------|
| Forward | 1.0ms | 3 |
| Backward | 2.0ms | 4 |
| Optimizer | 0.2ms | 3-4 |
| **Total** | **3.2ms** | **10-11** |

### Target Performance (Fused)
| Operation | Time | Kernels | Improvement |
|-----------|------|---------|-------------|
| Forward | 0.6ms | 1 | 1.7x |
| Backward | 1.0ms | 1 | 2.0x |
| Optimizer | 0.15ms | 1 | 1.3x |
| **Total** | **1.75ms** | **3** | **1.8x** |

### Memory Bandwidth Savings
| Pass | Current Reads | Current Writes | Fused Reads | Fused Writes | Savings |
|------|---------------|----------------|-------------|--------------|---------|
| Forward | 3 passes | 2 passes | 1 pass | 1 pass | 66% |
| Backward | 4 passes | 3 passes | 1 pass | 3 passes | 75% |

## 📚 References

- CUDA C++ Programming Guide: https://docs.nvidia.com/cuda/cuda-c-programming-guide/
- Kernel Fusion Techniques: https://developer.nvidia.com/blog/cuda-pro-tip-kernel-fusion/
- Phase 10 Plan: `LORA_GPU_PHASE10_PLAN.md`

### Fused Forward Pass Pseudocode
```cuda
__global__ void fused_lora_forward(
    float* input,  // [batch, in_features]
    float* B,      // [in_features, r]
    float* A,      // [r, out_features]
    float scaling,
    float* output  // [batch, out_features]
) {
    __shared__ float h_tile[TILE_SIZE][TILE_SIZE];
    
    // Compute h = input @ B in shared memory
    // Don't write to global memory
    
    // Compute output = h @ A * scaling
    // Write final result to global memory
}
```

### Fused Backward Pass Pseudocode
```cuda
__global__ void fused_lora_backward(
    float* grad_out,  // [batch, out_features]
    float* input,     // [batch, in_features]
    float* B,         // [in_features, r]
    float* A,         // [r, out_features]
    float scaling,
    float* grad_A,    // [r, out_features]
    float* grad_B,    // [in_features, r]
    float* grad_input // [batch, in_features]
) {
    // All gradients computed in single kernel
    // Use shared memory for intermediate results
    // Minimize global memory traffic
}
```

### Memory Access Patterns
- **Unfused**: Multiple passes over same data (poor cache utilization)
- **Fused**: Single pass over data (excellent cache utilization)
- **Result**: Significantly reduced memory bandwidth

---

**Status**: Not Started  
**Estimated Completion**: Week 8 of Phase 10  
**Expected Benefit**: 1.5-2x additional speedup (cumulative 75-100x vs CPU)
